vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,200 @@
+---
+title: "Data repository"
+author: "Thierry Onkelinx"
+date: "`r Sys.Date()`"
+output: rmarkdown::html_vignette
+vignette: >
+  %\VignetteIndexEntry{Data repository}
+  %\VignetteEngine{knitr::rmarkdown}
+  %\VignetteEncoding{UTF-8}
+---
+
+```{r setup, include = FALSE}
+library(knitr)
+opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+```
+
+## Rationale
+
+In some cases we want to place data under version control. Data refers here to the information which is stored in a rectangular format. The columns define variables, while the rows represent observations. There is no information in neither the order of the columns nor the order of the rows. We assume that the data is available a an R `data.frame`.
+
+Although git can store binary files, its real power and efficiency is with plain text files. R has several functions to read and write `data.frames` as plain text files. Main downside is the loss of meta data, especially in case of `factors`. A `factor` will be written as a `character` and hence lose all information on the levels. This can be important when the levels are a) not sorted alphabetical, b) not all levels have observations or c) the `factor` is ordered. Besides that, storing a `factor` as `character` is not efficient.
+
+Git stores the version history under the form of diffs: a list of lines which are deleted and a list of lines which are inserted at a specific line number in a file. We have to keep this in mind when storing data as plain text files in Git. The table below indicate what impact changing the information content has on the diff in Git. Note that changing observations requires a small diff, while changing variables results in a massive diff.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("remove 1 observation", "remove 1 line"),
+    c("add 1 observation", "add 1 line"),
+    c("update 1 observation", "remove 1 line and add 1 line"),
+    c("remove 1 variable", "remove all lines and add all lines"),
+    c("add 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the information content of the data")
+```
+
+In the next table we show the effect on the diff when making changes in the data which don't change in the information content in the data.
+
+```{r echo = FALSE}
+cases <- rbind(
+    c("move 1 observation", "remove 1 line and add 1 line"),
+    c("move 1 variable", "remove all lines and add all lines")
+)
+colnames(cases) <- c("Change in data", "Git diff")
+kable(cases, caption = "Changes in the data without changing the information content")
+```
+
+So in order to use Git a a performant and efficient version control system for data, we need to make sure to store the meta data and keep the diffs as small as possible.
```

Comment:
I would add something like to make the bridge with the net section:

'In the next section, the way `git2r` handles this will be explained'